synapseclient/models/mixins/storable_container.py

```diff
@@ -54,6 +55,42 @@ class StorableContainer(StorableContainerSynchronousProtocol):
     async def get_async(self, *, synapse_client: Optional[Synapse] = None) -> None:
         """Used to satisfy the usage in this mixin from the parent class."""
 
+    async def worker(
+        self,
+        queue: asyncio.Queue,
+        failure_strategy: FailureStrategy,
+        synapse_client: Synapse,
+    ) -> NoReturn:
+        """
+        Coroutine that will process the queue of work items. This will process the
+        work items until the queue is empty. This will be used to download files in
+        parallel.
+
+        Arguments:
+            queue: The queue of work items to process.
+            failure_strategy: Determines how to handle failures when retrieving items
+                out of the queue and an exception occurs.
+            synapse_client: The Synapse client to use to download the files.
+        """
+        while True:
+            # Get a "work item" out of the queue.
+            work_item = await queue.get()
+
+            try:
+                result = await work_item
+            except asyncio.CancelledError as ex:
+                raise ex
+            except Exception as ex:
+                result = ex
+
+            self._resolve_sync_from_synapse_result(
+                result=result,
+                failure_strategy=failure_strategy,
+                synapse_client=synapse_client,
+            )
+
+            queue.task_done()
```

Comment:
By taking this worker approach we are limiting the number of concurrent file transfers. Asyncio also provides a semaphore (https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) - But, this First in, First out queue seemed like the more intuitive approach to solving this problem.